test/org/centos/contra/Infra/Utils.groovy

```diff
@@ -0,0 +1,347 @@
+import spock.lang.*
+import com.homeaway.devtools.jenkins.testing.JenkinsPipelineSpecification
+
+import org.yaml.snakeyaml.Yaml
+
+public class UtilsSpec extends JenkinsPipelineSpecification {
+    @Shared
+    def config = new HashMap<String, ?>()
+
+    @Shared
+    def infraUtils
+
+    def setup() {
+        infraUtils = loadPipelineScriptForTest("../../src/org/centos/contra/Infra/Utils.groovy")
+        explicitlyMockPipelineStep('dir')
+        explicitlyMockPipelineStep('writeFile')
+        explicitlyMockPipelineStep('fileExists')
+        explicitlyMockPipelineStep('withEnv')
+        explicitlyMockPipelineStep('container')
+
+        explicitlyMockPipelineVariable('WORKSPACE')
+        explicitlyMockPipelineVariable('CREDENTIAL_FILE_NAME')
+        explicitlyMockPipelineVariable('SSH_FILE_NAME')
+    }
+
+    def setupSpec() {
+        Yaml yaml = new Yaml()
+        config = (Map<String, ?>) yaml.load(new FileReader('test/contra-sample.yml'))
+    }
+
+    def "create aws instances"() {
+        def aws = config.infra.provision.cloud.aws
+        def i = 1
+        when:
+        def instances = infraUtils.createAwsInstances(aws as HashMap)
+
+        then:
+        instances != null
+        // createAwsInstances modifies state so we need to modify it back
+        def base_name = aws.instances[0].name
```

Comment:
`Then` blocks really shouldn't have a whole lot of logic. I would say `instances != null` could actually go in a `expect` block. Also, what is the `modify it back`? Each test should live independently. So what happens in this test should not effect other tests and that should be taken care of in `setup` of `cleanup` methods. So i'm wondering how the state is being modified so you have to revert it to continue on?

Reply:
I think that comment is leftover from something I was experimenting with. I don't think it applies to anything in this test; I just missed it in cleanup and I'll remove it.